diff --git a/core/completer.py b/core/completer.py
index b1bc646..6ea0acf 100644
--- a/core/completer.py
+++ b/core/completer.py
@@ -74,12 +74,18 @@ import keyword
 import os
 import re
 import sys
+import numpy
+import types
+import repr as REPR
+try:
+    import funcsigs
+except ImportError:
+    pass
 
 from IPython.config.configurable import Configurable
 from IPython.core.error import TryNext
 from IPython.core.inputsplitter import ESC_MAGIC
 from IPython.utils import generics
-from IPython.utils import io
 from IPython.utils.dir2 import dir2
 from IPython.utils.process import arg_split
 from IPython.utils.py3compat import builtin_mod, string_types
@@ -89,6 +95,38 @@ from IPython.utils.traitlets import CBool, Enum
 # Globals
 #-----------------------------------------------------------------------------
 
+repr_ = REPR.Repr()
+repr_.maxdict = 0
+TYPES_LIST = tuple([getattr(types, a)
+                    for a in dir(types) if isinstance(getattr(types, a), type)])
+
+
+def _hassource(object):
+    """Check if object has source code worth fetching."""
+    if not hasattr(object, '__call__'):
+        return False
+    if inspect.ismodule(object):
+        if hasattr(object, '__file__'):
+            return True
+        return False
+    if inspect.isclass(object):
+        object = sys.modules.get(object.__module__)
+        if hasattr(object, '__file__'):
+            return True
+        return False
+    if inspect.ismethod(object):
+        object = object.im_func
+    if inspect.isfunction(object):
+        object = object.func_code
+    if inspect.istraceback(object):
+        object = object.tb_frame
+    if inspect.isframe(object):
+        object = object.f_code
+    if inspect.iscode(object):
+        return True
+    return False
+
+
 # Public API
 __all__ = ['Completer','IPCompleter']
 
@@ -97,6 +135,7 @@ if sys.platform == 'win32':
 else:
     PROTECTABLES = ' ()[]{}?=\\|;:\'#*"^&'
 
+
 #-----------------------------------------------------------------------------
 # Main functions and classes
 #-----------------------------------------------------------------------------
@@ -426,6 +465,64 @@ def get__all__entries(obj):
     return [w for w in words if isinstance(w, string_types)]
 
 
+def match_dict_keys(keys, prefix):
+    """Used by dict_key_matches, matching the prefix to a list of keys"""
+    if not prefix:
+        return None, 0, [repr(k) for k in keys
+                      if isinstance(k, (string_types, bytes))]
+    quote_match = re.search('["\']', prefix)
+    quote = quote_match.group()
+    try:
+        prefix_str = eval(prefix + quote, {})
+    except Exception:
+        return None, 0, []
+    
+    token_match = re.search(r'\w*$', prefix, re.UNICODE)
+    token_start = token_match.start()
+    token_prefix = token_match.group()
+
+    # TODO: support bytes in Py3k
+    matched = []
+    for key in keys:
+        try:
+            if not key.startswith(prefix_str):
+                continue
+        except (AttributeError, TypeError, UnicodeError):
+            # Python 3+ TypeError on b'a'.startswith('a') or vice-versa
+            continue
+
+        # reformat remainder of key to begin with prefix
+        rem = key[len(prefix_str):]
+        # force repr wrapped in '
+        rem_repr = repr(rem + '"')
+        if rem_repr.startswith('u') and prefix[0] not in 'uU':
+            # Found key is unicode, but prefix is Py2 string.
+            # Therefore attempt to interpret key as string.
+            try:
+                rem_repr = repr(rem.encode('ascii') + '"')
+            except UnicodeEncodeError:
+                continue
+
+        rem_repr = rem_repr[1 + rem_repr.index("'"):-2]
+        if quote == '"':
+            # The entered prefix is quoted with ",
+            # but the match is quoted with '.
+            # A contained " hence needs escaping for comparison:
+            rem_repr = rem_repr.replace('"', '\\"')
+
+        # then reinsert prefix from start of token
+        matched.append('%s%s' % (token_prefix, rem_repr))
+    return quote, token_start, matched
+
+
+def _safe_isinstance(obj, module, class_name):
+    """Checks if obj is an instance of module.class_name if loaded
+    """
+    return (module in sys.modules and
+            isinstance(obj, getattr(__import__(module), class_name)))
+
+
+
 class IPCompleter(Completer):
     """Extension of the completer class with IPython-specific features"""
 
@@ -538,6 +635,7 @@ class IPCompleter(Completer):
                          self.file_matches,
                          self.magic_matches,
                          self.python_func_kw_matches,
+                         self.dict_key_matches,
                          ]
 
     def all_completions(self, text):
@@ -804,6 +902,108 @@ class IPCompleter(Completer):
                     argMatches.append("%s=" %namedArg)
         return argMatches
 
+    def dict_key_matches(self, text):
+        "Match string keys in a dictionary, after e.g. 'foo[' "
+        def get_keys(obj):
+            # Only allow completion for known in-memory dict-like types
+            if isinstance(obj, dict) or\
+               _safe_isinstance(obj, 'attrdict', 'AttrDict') or\
+               _safe_isinstance(obj, 'pandas', 'DataFrame'):
+                try:
+                    return list(obj.keys())
+                except Exception:
+                    return []
+            elif _safe_isinstance(obj, 'numpy', 'ndarray'):
+                return obj.dtype.names or []
+            return []
+
+        try:
+            regexps = self.__dict_key_regexps
+        except AttributeError:
+            dict_key_re_fmt = r'''(?x)
+            (  # match dict-referring expression wrt greedy setting
+                %s
+            )
+            \[   # open bracket
+            \s*  # and optional whitespace
+            ([uUbB]?  # string prefix (r not handled)
+                (?:   # unclosed string
+                    '(?:[^']|(?<!\\)\\')*
+                |
+                    "(?:[^"]|(?<!\\)\\")*
+                )
+            )?
+            $
+            '''
+            regexps = self.__dict_key_regexps = {
+                False: re.compile(dict_key_re_fmt % '''
+                                  # identifiers separated by .
+                                  (?!\d)\w+
+                                  (?:\.(?!\d)\w+)*
+                                  '''),
+                True: re.compile(dict_key_re_fmt % '''
+                                 [^ =\r\n]+
+                                 ''')
+            }
+
+        match = regexps[self.greedy].search(self.text_until_cursor)
+        if match is None:
+            return []
+
+        expr, prefix = match.groups()
+        try:
+            obj = eval(expr, self.namespace)
+        except Exception:
+            try:
+                obj = eval(expr, self.global_namespace)
+            except Exception:
+                return []
+
+        keys = get_keys(obj)
+        if not keys:
+            return keys
+        closing_quote, token_offset, matches = match_dict_keys(keys, prefix)
+        if not matches:
+            return matches
+        
+        # get the cursor position of
+        # - the text being completed
+        # - the start of the key text
+        # - the start of the completion
+        text_start = len(self.text_until_cursor) - len(text)
+        if prefix:
+            key_start = match.start(2)
+            completion_start = key_start + token_offset
+        else:
+            key_start = completion_start = match.end()
+        
+        # grab the leading prefix, to make sure all completions start with `text`
+        if text_start > key_start:
+            leading = ''
+        else:
+            leading = text[:completion_start-text_start]
+        
+        # the index of the `[` character
+        bracket_idx = match.end(1)
+
+        # append closing quote and bracket as appropriate
+        # this is *not* appropriate if the opening quote or bracket is outside
+        # the text given to this method
+        suf = ''
+        continuation = self.line_buffer[len(self.text_until_cursor):]
+        if key_start > text_start and closing_quote:
+            # quotes were opened inside text, maybe close them
+            if continuation.startswith(closing_quote):
+                continuation = continuation[len(closing_quote):]
+            else:
+                suf += closing_quote
+        if bracket_idx > text_start:
+            # brackets were opened inside text, maybe close them
+            if not continuation.startswith(']'):
+                suf += ']'
+        
+        return [leading + k + suf for k in matches]
+
     def dispatch_custom_completer(self, text):
         #io.rprint("Custom! '%s' %s" % (text, self.custom_completers)) # dbg
         line = self.line_buffer
@@ -931,8 +1131,75 @@ class IPCompleter(Completer):
         # use penalize_magics_key to put magics after variables with same name
         self.matches = sorted(set(self.matches), key=penalize_magics_key)
 
+        matches = []
+        for i, m in enumerate(self.matches):
+            try:
+                obj = eval(m, self.namespace)
+            except Exception:
+                try:
+                    obj = eval(m, self.global_namespace)
+                except Exception:
+                    matches.append(m)
+                    continue
+
+            try:
+                assert(isinstance(obj, numpy.ndarray))
+                matches.append(m + '\0ndarray: ' + str(getattr(obj, 'shape')))
+            except (AssertionError, AttributeError, NameError):
+                try:
+                    matches.append(
+                        m + '\0' + '%s: %s.%s' %
+                        (type(obj).__name__,
+                         obj.__module__,
+                         obj.__name__))
+                except (AttributeError, KeyError):
+                    if type(obj) in TYPES_LIST:
+                        matches.append(
+                            m + '\0' + '%s: %s' %
+                            (type(obj).__name__,
+                             repr_.repr(obj)[: 50]))
+                    else:
+                        try:
+                            matches.append(
+                                m + '\0' +
+                                type(obj).__module__ + '.' +
+                                type(obj).__name__)
+                        except AttributeError:
+                            matches.append(m + '\0' + type(obj).__name__)
+
+            info = ''
+            if _hassource(obj):
+                try:
+                    info += obj.__name__ + unicode(
+                        funcsigs.signature(
+                            obj)) + '\n\n'
+                except (AttributeError, KeyError, NameError, TypeError,
+                        ValueError):
+                    try:
+                        source = inspect.getsource(obj)
+                    except (IOError, TypeError):
+                        pass
+                    else:
+                        def_ = re.sub(r'\s+', ' ', re.split(
+                            r'\)\s*:\s*\n', source)[0]) + ')\n\n'
+                        if def_.startswith('def '):
+                            info += def_[4:]
+                        elif def_.startswith('class '):
+                            info += def_[6:]
+
+            try:
+                info += inspect.getdoc(obj)
+            except TypeError:
+                pass
+
+            matches[i] += 'CALLSIG' + info if info else ''
+
+            matches[i] = matches[i].replace(
+                'builtin_function_or_method',
+                'builtin')
+
         #io.rprint('COMP TEXT, MATCHES: %r, %r' % (text, self.matches)) # dbg
-        return text, self.matches
+        return text, matches
 
     def rlcomplete(self, text, state):
         """Return the state-th possible completion for 'text'.
diff --git a/kernel/zmq/ipkernel.py b/kernel/zmq/ipkernel.py
index f76d506..58c8209 100644
--- a/kernel/zmq/ipkernel.py
+++ b/kernel/zmq/ipkernel.py
@@ -477,8 +477,16 @@ class Kernel(Configurable):
 
     def complete_request(self, stream, ident, parent):
         txt, matches = self._complete(parent)
+        metadata = []
+        for i, m in enumerate(matches):
+            if '\0' in m:
+                metadata.append(m.partition('\0')[-1])
+                matches[i] = m.partition('\0')[0]
+            else:
+                metadata.append('')
         matches = {'matches' : matches,
                    'matched_text' : txt,
+                   'metadata' : metadata,
                    'status' : 'ok'}
         matches = json_clean(matches)
         completion_msg = self.session.send(stream, 'complete_reply',
