diff --git a/src/ex_getln.c b/src/ex_getln.c
--- a/src/ex_getln.c
+++ b/src/ex_getln.c
@@ -6354,6 +6354,10 @@
     return (int)ccline.cmdbuff[ccline.cmdpos + offset];
 }
 #endif
+#ifdef FEAT_FOLDING
+    int                   save_KeyTyped;
+#endif
+
 
 #if defined(FEAT_CMDWIN) || defined(PROTO)
 /*
@@ -6527,9 +6531,20 @@
     save_KeyTyped = KeyTyped;
 #  endif
 
+
+# ifdef FEAT_FOLDING
+    save_KeyTyped = KeyTyped;
+# endif
+
     /* Trigger CmdwinLeave autocommands. */
     apply_autocmds(EVENT_CMDWINLEAVE, typestr, typestr, FALSE, curbuf);
 
+# ifdef FEAT_FOLDING
+    /* Restore KeyTyped in case it is modified by autocommands */
+    KeyTyped = save_KeyTyped;
+# endif
+
+
 #  ifdef FEAT_FOLDING
     /* Restore KeyTyped in case it is modified by autocommands */
     KeyTyped = save_KeyTyped;
