diff --git a/src/testdir/Makefile b/src/testdir/Makefile
--- a/src/testdir/Makefile
+++ b/src/testdir/Makefile
@@ -32,7 +32,7 @@
 		test89.out test90.out test91.out test92.out test93.out \
 		test94.out test95.out test96.out test97.out test98.out \
 		test99.out test100.out test101.out test102.out test103.out \
-		test104.out test105.out test106.out test107.out \
+		test104.out test105.out test106.out test107.out test108.out \
 		test_autoformat_join.out \
 		test_breakindent.out \
 		test_listlbr.out \
diff --git a/src/testdir/test108.in b/src/testdir/test108.in
new file mode 100644
--- /dev/null
+++ b/src/testdir/test108.in
@@ -0,0 +1,22 @@
+Test changelist position after splitting window
+Set 'undolevels' to make changelist for sourced file
+
+STARTTEST
+:so small.vim
+Gkylp:set ul=100
+Gylp:set ul=100
+gg
+:vsplit
+:try
+:  normal g;
+:  normal ggVGcpass
+:catch
+:  normal ggVGcfail
+:finally
+:  %w! test.out
+:endtry
+:qa!
+ENDTEST
+
+1
+2
diff --git a/src/testdir/test108.ok b/src/testdir/test108.ok
new file mode 100644
--- /dev/null
+++ b/src/testdir/test108.ok
@@ -0,0 +1,1 @@
+pass
diff --git a/src/window.c b/src/window.c
--- a/src/window.c
+++ b/src/window.c
@@ -1178,6 +1178,11 @@
 	    p_wh = size;
     }
 
+#ifdef FEAT_JUMPLIST
+    /* Keep same changelist position in new window */
+    wp->w_changelistidx = oldwin->w_changelistidx;
+#endif
+
     /*
      * make the new window the current window
      */
