diff --git a/src/ui.c b/src/ui.c
--- a/src/ui.c
+++ b/src/ui.c
@@ -1396,7 +1396,7 @@
 clip_gen_own_selection(cbd)
     VimClipboard	*cbd;
 {
-#ifdef FEAT_XCLIPBOARD
+#if defined(FEAT_XCLIPBOARD) && !defined(MACOS_X_UNIX)
 # ifdef FEAT_GUI
     if (gui.in_use)
 	return clip_mch_own_selection(cbd);
@@ -1412,7 +1412,7 @@
 clip_gen_lose_selection(cbd)
     VimClipboard	*cbd;
 {
-#ifdef FEAT_XCLIPBOARD
+#if defined(FEAT_XCLIPBOARD) && !defined(MACOS_X_UNIX)
 # ifdef FEAT_GUI
     if (gui.in_use)
 	clip_mch_lose_selection(cbd);
@@ -1428,7 +1428,7 @@
 clip_gen_set_selection(cbd)
     VimClipboard	*cbd;
 {
-#ifdef FEAT_XCLIPBOARD
+#if defined(FEAT_XCLIPBOARD) && !defined(MACOS_X_UNIX)
 # ifdef FEAT_GUI
     if (gui.in_use)
 	clip_mch_set_selection(cbd);
@@ -1444,7 +1444,7 @@
 clip_gen_request_selection(cbd)
     VimClipboard	*cbd;
 {
-#ifdef FEAT_XCLIPBOARD
+#if defined(FEAT_XCLIPBOARD) && !defined(MACOS_X_UNIX)
 # ifdef FEAT_GUI
     if (gui.in_use)
 	clip_mch_request_selection(cbd);
@@ -1460,7 +1460,7 @@
 clip_gen_owner_exists(cbd)
     VimClipboard	*cbd UNUSED;
 {
-#ifdef FEAT_XCLIPBOARD
+#if defined(FEAT_XCLIPBOARD) && !defined(MACOS_X_UNIX)
 # ifdef FEAT_GUI_GTK
     if (gui.in_use)
 	return clip_gtk_owner_exists(cbd);
diff --git a/src/version.c b/src/version.c
--- a/src/version.c
+++ b/src/version.c
@@ -60,6 +60,7 @@
 
 static char *(features[]) =
 {
+	"+mac_clipboard",
 #ifdef HAVE_ACL
 	"+acl",
 #else
