diff --git a/src/screen.c b/src/screen.c
index a72bffa..8d70e2f 100644
--- a/src/screen.c
+++ b/src/screen.c
@@ -8960,9 +8960,12 @@ can_clear(char_u *p)
 		|| gui.in_use
 #endif
 #ifdef FEAT_TERMGUICOLORS
-		|| (p_tgc && cterm_normal_bg_gui_color != (long_u)INVALCOLOR)
+		|| (p_tgc && cterm_normal_bg_gui_color == (long_u)INVALCOLOR)
+		|| (!p_tgc && cterm_normal_bg_color == 0)
+#else
+		|| cterm_normal_bg_color == 0
 #endif
-		|| cterm_normal_bg_color == 0 || *T_UT != NUL));
+		|| *T_UT != NUL));
 }
 
 /*
