diff --git a/src/edit.c b/src/edit.c
--- a/src/edit.c
+++ b/src/edit.c
@@ -4596,6 +4596,8 @@
     /* TODO: is this sufficient for redrawing?  Redrawing everything causes
      * flicker, thus we can't do that. */
     changed_cline_bef_curs();
+    /* Clean up deleted characters */
+    screen_start();
 }
 
 /* Insert the new text being completed. */
diff --git a/src/version.c b/src/version.c
--- a/src/version.c
+++ b/src/version.c
@@ -729,6 +729,7 @@
 	"-xpm",
 # endif
 #endif
+	"+ins_compl_del",
 	NULL
 };
 
